torchtune/dev/rl/workers/parameter_servers/vllm.py

```diff
@@ -0,0 +1,128 @@
+class vLLMParameterServer(RemoteWeightUpdaterBase):
+    def __init__(self, vllm_master_addresses, vllm_master_ports, env_vars):
+        log.info("in param server init")
+        super().__init__()
+        self.vllm_master_addresses = vllm_master_addresses
+        self.vllm_master_ports = vllm_master_ports
+        self.vllm_comm_groups = dict()
+        self.vllm_weight_versions = dict()
+        self.vllm_worker_handles = dict()
+
+        import os
+
+        import torch
+        import torch.distributed
+
+        torch.cuda.set_device(torch.device("cuda", 0))
+
+        for var in env_vars:
+            os.environ[var] = str(env_vars[var])
+
+        if not torch.distributed.is_initialized():
+            torch.distributed.init_process_group(
+                backend="nccl", device_id=torch.device("cuda:0")
+            )
+
+        self.rank = int(os.environ["RANK"])
+        self.world_size = int(os.environ["WORLD_SIZE"])
+        assert self.rank == self.world_size - 1
+
+        # FIXME: why this hang even when I pass use_local_synchronization=False in the other one??
+        # self.fsdp_group = torch.distributed.new_group(ranks=list(range(self.world_size - 1)))
+
+    def register_collector(self, worker_id, handle):
+        self.vllm_worker_handles[worker_id] = handle
+        log.info(f"registered collector {worker_id=}")
+
+    def register_model_metadata(self, model_metadata):
+        self.model_metadata = model_metadata
+        self.state_dict = dict()
+        for k, (dtype, shape) in model_metadata.items():
+            self.state_dict[k] = torch.zeros(shape, dtype=dtype, device="cuda")
+        self.state_dict_lock = rwlock.RWLockFairD()
+        self.version = 0
+        self.version_tensor = torch.tensor([0], device="cuda")
+
+    def acquire_state_dict_lock(self):
+        self.write_lock = self.state_dict_lock.gen_wlock()
+        self.write_lock.acquire()
+
+    def release_state_dict_lock(self):
+        self.version += 1
+        self.version_tensor += 1
+        torch.cuda.synchronize()
+        self.write_lock.release()
+
+    def all_worker_ids(self):
+        return [i for i in range(len(self.collector._remote_collectors))]
+
+    def _get_server_weights(self):
+        return self.state_dict
+
+    def _maybe_map_weights(self, server_weights):
+        return server_weights
+
+    def _skip_update(self, worker_id):
+        if self.version == 0:
+            return True
+        if worker_id not in self.vllm_weight_versions:
+            return False
+        if self.vllm_weight_versions[worker_id] == self.version:
+            log.info(
+                f"skipping update for {worker_id=}, {self.version=}, {self.vllm_weight_versions[worker_id]=}"
+            )
+            return True
+        return False
+
+    def _init_model_update_group(self, worker_id):
+        # here again, I want to grab the tp size from the vLLM worker... :(
+        # llm.llm_engine.parallel_config.tensor_parallel_size
```

Comment:
Hmmm I think #33 might have been lost in the refactor, I can def fix this after this PR lands though!!

Reply:
No worries, that was before rebase! Let me update now. It's not working tho